index.d.ts

```diff
@@ -0,0 +1,252 @@
+declare module 'ethers' {
+
+  class Wallet {
+    public constructor(privateKey: SigningKey | string, provider: providers.Provider)
+
+    static createRandom(options?: any): Wallet
+    static verifyMessage(message: string, signature: string): string
+    static parseTransaction(rawTransaction: any): Transaction
+    static isEncryptedWallet(json: JSON): boolean
+    static fromEncryptedWallet(json: JSON, password: string, progressCallback?: Function): Promise<Wallet>
+    static fromMnemonic(mnemonic: string, path: string): Wallet
+    static fromBrainWallet(username: string, password: string, progressCallback?: Function): Promise<Wallet>
+
+    readonly privateKey: string
+    readonly address: string
+
+    provider?: providers.Provider
+    defaultGasLimit: number
+
+    sign(transaction: Transaction): utils.RLP
+    signMessage(message: string): string
+    encrypt(password: string, options: any, progressCallback?: Function): Promise<any>
+
+    getAddress(): string
+    getBalance(blockTag: string): Promise<number>
+    getTransactionCount(blockTag: string): Promise<number>
+    estimateGas(transaction: Transaction): Promise<number>
+    sendTransaction(transaction: Transaction): Promise<Transaction>
+    send(addressOrENSName: string, amountWei: utils.BigNumber, options?: any): Promise<Transaction>
+  }
+
+  class SigningKey {
+    public constructor(privateKey: any)
+    static recover(digest: any, r: string, s: string, recoveryParam: any): SigningKey
+    static getPublicKey(value: any, compressed?: boolean): string
+    static publicKeyToAddress(publicKey: any): string
+
+    readonly privateKey: string
+    readonly publicKey: string
+    readonly address: string
+
+    signDigest(digest: any): any
+  }
+
+  class Contract {
+    public constructor(addressOrENSName: string, contractInterface: string | object | Interface, signerOrProvider: any)
+
+    readonly address: string
+    readonly contractInterface: Interface
+    readonly signer: any
+    readonly provider: providers.JsonRpcProvider
+    readonly estimate: any
+    readonly functions: any
+    readonly events: any
+
+    connect(signerOrProvider: any): Contract
+    getDeployTransaction(bytecode: any, contractInterface: Interface): any
```

Comment:
This should be a static method on `Contract`.  Also, the parameters and return type are incorrect:
```typescript
static getDeployTransaction(bytecode: string, contractInterface: string, ...constructorArguments: Array<any>): Transaction
```

Reply:
Note to ethers maintainers, it really feels like `contractInterface` should be a well defined object rather than a JSON string.  This is what I use:
```typescript
	export type Primitive = 'uint8' | 'uint64' | 'uint256' | 'bool' | 'string' | 'address' | 'bytes20' | 'bytes32' | 'bytes' | 'int256' | 'address[]' | 'uint256[]' | 'bytes32[]';

	export interface AbiParameter {
		name: string,
		type: Primitive,
	}

	export interface AbiEventParameter extends AbiParameter {
		indexed: boolean,
	}

	export interface AbiFunction {
		name: string,
		type: 'function' | 'constructor' | 'fallback',
		stateMutability: 'pure' | 'view' | 'payable' | 'nonpayable',
		constant: boolean,
		payable: boolean,
		inputs: Array<AbiParameter>,
		outputs: Array<AbiParameter>,
	}

	export interface AbiEvent {
		name: string,
		type: 'event',
		inputs: Array<AbiEventParameter>,
		anonymous: boolean,
	}

	export type Abi = Array<AbiFunction | AbiEvent>;
```